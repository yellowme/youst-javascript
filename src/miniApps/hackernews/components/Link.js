import React from "react";

import { timeDifferenceForDate } from "../utils";

import Text from "../../../core/components/Text";

export default function Link({
  id,
  index,
  description,
  url,
  votes,
  postedBy,
  createdAt,
  onVote
}) {
  const handleClick = () => onVote(id);

  return (
    <div className="flex mt2 items-start">
      <div className="flex items-center">
        <span className="gray">{index + 1}.</span>
        <div onClick={handleClick} className="ml1 gray f11">
          â–²
        </div>
      </div>
      <div className="ml1">
        <Text>
          {description} ({url})
        </Text>
        <div className="f6 lh-copy gray">
          {votes} votes | by {postedBy ? postedBy.name : "Unknown"}{" "}
          {timeDifferenceForDate(createdAt)}
        </div>
      </div>
    </div>
  );
}
