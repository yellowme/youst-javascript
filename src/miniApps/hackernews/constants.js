export const AUTH_TOKEN = "AUTH_TOKEN_MINI_APP_HACKER_NEWS";
export const USER_ID = "USER_ID_MINI_APP_HACKER_NEWS";
export const LINKS_PER_PAGE = 5;
export const MINI_APP_BASE_ROUTE = '/mini-apps/hacker-news'
