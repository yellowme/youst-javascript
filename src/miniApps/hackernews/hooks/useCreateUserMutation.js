import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";

export default function useCreateUserMutation() {
  const [mutate] = useMutation(CREATE_USER_MUTATION);

  function createUser({ name, email, password }) {
    return mutate({
      variables: {
        name,
        authProvider: {
          email: {
            email,
            password
          }
        }
      }
    });
  }

  return createUser;
}

export const CREATE_USER_MUTATION = gql`
  mutation CreateUserMutation(
    $name: String!
    $authProvider: AuthProviderSignupData!
  ) {
    createUser(name: $name, authProvider: $authProvider) {
      id
    }
  }
`;
