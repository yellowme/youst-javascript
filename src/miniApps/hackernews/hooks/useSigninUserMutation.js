import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";

export default function useSigninUserMutation() {
  const [mutate] = useMutation(SIGNIN_USER_MUTATION);

  function signinUser({ email, password }) {
    return mutate({
      variables: { email: { email, password } }
    });
  }

  return signinUser;
}

export const SIGNIN_USER_MUTATION = gql`
  mutation SigninUserMutation($email: AUTH_PROVIDER_EMAIL) {
    signinUser(email: $email) {
      token
      user {
        id
      }
    }
  }
`;
