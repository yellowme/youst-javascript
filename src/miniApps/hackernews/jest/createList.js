export default function createList(length, factory) {
  return Array.from({ length }).map(factory);
}
