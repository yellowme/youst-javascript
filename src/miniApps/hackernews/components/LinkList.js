import React from "react";
import { useQuery, useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";

import { LINKS_PER_PAGE, AUTH_TOKEN, USER_ID } from "../constants";
import Link from "./Link";

export default function LinkList({ location, match, history }) {
  const isNewPage = location.pathname.includes("new");
  const page = parseInt(match.params.page, 10);

  const [mutate] = useMutation(CREATE_VOTE_MUTATION);
  const { loading, error, data } = useQuery(ALL_LINKS_QUERY, {
    variables: getQueryVariables()
  });

  function handleVoteMutation(linkId) {
    const hasToken = localStorage.getItem(AUTH_TOKEN);
    const userId = localStorage.getItem(USER_ID);
    if (!hasToken) return history.push(`/mini-apps/hacker-news/login`);

    return mutate({
      mutation: CREATE_VOTE_MUTATION,
      variables: { linkId, userId },
      update: (store, { data: { createVote } }) => {
        let skip = 0;
        let first = 100;
        let orderBy = null;

        if (isNewPage) {
          skip = (page - 1) * LINKS_PER_PAGE;
          first = LINKS_PER_PAGE;
          orderBy = "createdAt_DESC";
        }

        const data = store.readQuery({
          query: ALL_LINKS_QUERY,
          variables: { first, skip, orderBy }
        });

        const votedLink = data.allLinks.find(link => link.id === linkId);
        votedLink.votes.push(createVote);
        store.writeQuery({ query: ALL_LINKS_QUERY, data });
      }
    });
  }

  function getQueryVariables() {
    const page = parseInt(match.params.page, 10);

    let skip = 0;
    let first = 100;
    let orderBy = null;

    if (isNewPage) {
      skip = (page - 1) * LINKS_PER_PAGE;
      first = LINKS_PER_PAGE;
      orderBy = "createdAt_DESC";
    }

    return { first, skip, orderBy };
  }

  function getLinksToRender() {
    if (isNewPage) return data.allLinks;
    // Prevent in place mutation
    const rankedLinks = [...data.allLinks].sort(
      (l1, l2) => l2.votes.length - l1.votes.length
    );

    return rankedLinks;
  }

  function goToPreviousPage() {
    const page = parseInt(match.params.page, 10);
    if (page > 1) {
      const previousPage = page - 1;
      history.push(`/new/${previousPage}`);
    }
  }

  function goToNextPage() {
    const page = parseInt(match.params.page, 10);
    if (page <= data.allLinks.length / LINKS_PER_PAGE) {
      const nextPage = page + 1;
      history.push(`/new/${nextPage}`);
    }
  }

  if (loading) return <div>Loading...</div>;
  if (error) return <div>Error!</div>;

  return (
    <div>
      {getLinksToRender().map((link, index) => (
        <Link
          id={link.id}
          key={link.id}
          url={link.url}
          index={index}
          description={link.description}
          votes={link.votes.length}
          postedBy={link.postedBy}
          createdAt={link.createdAt}
          onVote={handleVoteMutation}
        />
      ))}
      {isNewPage && (
        <div className="flex ml4 mv3 gray">
          <div className="pointer mr2" onClick={goToPreviousPage}>
            Previous
          </div>
          <div className="pointer" onClick={goToNextPage}>
            Next
          </div>
        </div>
      )}
    </div>
  );
}

export const ALL_LINKS_QUERY = gql`
  query AllLinksQuery($first: Int, $skip: Int, $orderBy: LinkOrderBy) {
    allLinks(first: $first, skip: $skip, orderBy: $orderBy) {
      id
      url
      description
      createdAt
      postedBy {
        id
        name
      }
      votes {
        id
        user {
          id
        }
      }
    }
  }
`;

export const CREATE_VOTE_MUTATION = gql`
  mutation CreateVoteMutation($linkId: ID, $userId: ID) {
    createVote(linkId: $linkId, userId: $userId) {
      id
      user {
        id
      }
    }
  }
`;
