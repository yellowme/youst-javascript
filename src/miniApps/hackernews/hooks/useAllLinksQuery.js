import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";

import { LINKS_PER_PAGE } from "../constants";

export default function useAllLinksQuery({ page, isNewPage }) {
  const { data = { allLinks: [] }, error, loading } = useQuery(
    ALL_LINKS_QUERY,
    {
      variables: getQueryVariables()
    }
  );

  function getQueryVariables() {
    let skip = 0;
    let first = 100;
    let orderBy = null;

    if (isNewPage) {
      skip = (page - 1) * LINKS_PER_PAGE;
      first = LINKS_PER_PAGE;
      orderBy = "createdAt_DESC";
    }

    return { first, skip, orderBy };
  }

  return [data.allLinks, loading, error];
}

export const ALL_LINKS_QUERY = gql`
  query AllLinksQuery($first: Int, $skip: Int, $orderBy: LinkOrderBy) {
    allLinks(first: $first, skip: $skip, orderBy: $orderBy) {
      id
      url
      description
      createdAt
      postedBy {
        id
        name
      }
      votes {
        id
        user {
          id
        }
      }
    }
  }
`;
