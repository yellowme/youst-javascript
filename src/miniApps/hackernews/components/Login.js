import React, { useState } from "react";
import { useFormik } from "formik";

import { MINI_APP_BASE_ROUTE } from "../constants";
import useAuthenticatedUser from "../hooks/useUser";
import useSigninUserMutation from "../hooks/useSigninUserMutation";
import useCreateUserMutation from "../hooks/useCreateUserMutation";

export default function Login({ history }) {
  const { login } = useAuthenticatedUser();
  const createUser = useCreateUserMutation();
  const signinUser = useSigninUserMutation();
  const [displayLoginForm, setDisplayForm] = useState(true);

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
      name: ""
    },
    onSubmit: handleSubmit
  });

  async function handleSubmit(variables) {
    try {
      if (!displayLoginForm) {
        await handleCreateUserMutation(variables);
      }

      const { data } = await handleSigninUserMutation(variables);
      const { token: authToken, user } = data.signinUser;
      login({ authToken, userId: user.id });
      history.push(MINI_APP_BASE_ROUTE);
    } catch (error) {
      console.log(error);
    }
  }

  function handleCreateUserMutation(variables) {
    return createUser({
      name: variables.name,
      email: variables.email,
      password: variables.password
    });
  }

  function handleSigninUserMutation(variables) {
    return signinUser({
      email: variables.email,
      password: variables.password
    });
  }

  return (
    <form onSubmit={formik.handleSubmit}>
      <h4 className="mv3">{displayLoginForm ? "Login" : "Sign Up"}</h4>
      <div className="flex flex-column">
        {!displayLoginForm && (
          <input
            name="name"
            onChange={formik.handleChange}
            value={formik.values.name}
            type="text"
            placeholder="Your name"
          />
        )}
        <input
          name="email"
          onChange={formik.handleChange}
          value={formik.values.email}
          type="text"
          placeholder="Your email address"
        />
        <input
          name="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          type="password"
          placeholder="Choose a safe password"
        />
      </div>
      <div className="flex mt3">
        <div className="pointer mr2 button" onClick={formik.handleSubmit}>
          {displayLoginForm ? "login" : "create account"}
        </div>
        <div
          className="pointer button"
          onClick={() => setDisplayForm(!displayLoginForm)}
        >
          {displayLoginForm
            ? "need to create an account?"
            : "already have an account?"}
        </div>
      </div>
    </form>
  );
}
