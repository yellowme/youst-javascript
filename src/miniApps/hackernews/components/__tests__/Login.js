import React from "react";
import renderer from "react-test-renderer";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter, Redirect, Route } from "react-router-dom";
import faker from "faker";
import wait from "waait";

import { CREATE_USER_MUTATION } from "../../hooks/useCreateUserMutation";
import { SIGNIN_USER_MUTATION } from "../../hooks/useSigninUserMutation";
import { ALL_LINKS_QUERY } from "../../hooks/useAllLinksQuery";
import { MINI_APP_BASE_ROUTE, LINKS_PER_PAGE } from "../../constants";
import createUser from "../../jest/factories/users";
import createLink from "../../jest/factories/link";
import LinkList from "../LinkList";
import Login from "../Login";
import Link from "../Link";
import createList from "../../jest/createList";

function TestLogin({ mocks }) {
  return (
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter initialEntries={[`${MINI_APP_BASE_ROUTE}/login`]}>
        <Route
          exact
          path={MINI_APP_BASE_ROUTE}
          render={() => <Redirect to={`${MINI_APP_BASE_ROUTE}/new/1`} />}
        />
        <Route
          exact
          path={`${MINI_APP_BASE_ROUTE}/new/:page`}
          component={LinkList}
        />
        <Route exact path={`${MINI_APP_BASE_ROUTE}/login`} component={Login} />
      </MemoryRouter>
    </MockedProvider>
  );
}

test("signup and login", async () => {
  const allLinks = createList(LINKS_PER_PAGE, createLink);
  const token = faker.lorem.word();
  const expectedUser = createUser({
    id: faker.random.uuid(),
    email: faker.internet.email(),
    password: faker.lorem.word()
  });

  const mocks = [
    {
      request: {
        query: CREATE_USER_MUTATION,
        variables: {
          name: expectedUser.name,
          authProvider: {
            email: {
              email: expectedUser.email,
              password: expectedUser.password
            }
          }
        }
      },
      result: {
        data: {
          createUser: {
            id: expectedUser.id
          }
        }
      }
    },
    {
      request: {
        query: SIGNIN_USER_MUTATION,
        variables: {
          email: {
            email: expectedUser.email,
            password: expectedUser.password
          }
        }
      },
      result: {
        data: {
          signinUser: {
            token,
            user: {
              id: expectedUser.id
            }
          }
        }
      }
    },
    {
      request: {
        query: ALL_LINKS_QUERY,
        variables: {
          first: LINKS_PER_PAGE,
          skip: 0,
          orderBy: "createdAt_DESC"
        }
      },
      result: {
        data: {
          allLinks
        }
      }
    }
  ];

  const component = renderer.create(<TestLogin mocks={mocks} />);

  // toggle login input to singup input
  const displayLoginFormButton = component.root.findByProps({
    className: "pointer button"
  });

  // Login Form
  expect(component).toMatchSnapshot();

  renderer.act(() => {
    displayLoginFormButton.props.onClick();
  });

  // Signup Form
  expect(component).toMatchSnapshot();

  const form = component.root.findByType("form");
  const nameInput = component.root.findByProps({ name: "name" });
  const emailInput = component.root.findByProps({ name: "email" });
  const passwordInput = component.root.findByProps({ name: "password" });

  // fill form
  await renderer.act(async () => {
    nameInput.props.onChange(
      createInputEvent({ name: "name", value: expectedUser.name })
    );

    emailInput.props.onChange(
      createInputEvent({ name: "email", value: expectedUser.email })
    );

    passwordInput.props.onChange(
      createInputEvent({ name: "password", value: expectedUser.password })
    );

    await form.props.onSubmit();
  });

  await renderer.act(async () => {
    await wait(0); // Wait to LinkList query resolves
  });

  // FIXME:
  // const links = component.root.findAllByType(Link);
  // expect(links.length).toBe(allLinks.length);
});

function createInputEvent({ name, value }) {
  return {
    target: { name, value }
  };
}
