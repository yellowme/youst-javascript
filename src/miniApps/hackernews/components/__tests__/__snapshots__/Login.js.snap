// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup and login 1`] = `
<form
  onSubmit={[Function]}
>
  <h4
    className="mv3"
  >
    Login
  </h4>
  <div
    className="flex flex-column"
  >
    <input
      name="email"
      onChange={[Function]}
      placeholder="Your email address"
      type="text"
      value=""
    />
    <input
      name="password"
      onChange={[Function]}
      placeholder="Choose a safe password"
      type="password"
      value=""
    />
  </div>
  <div
    className="flex mt3"
  >
    <div
      className="pointer mr2 button"
      onClick={[Function]}
    >
      login
    </div>
    <div
      className="pointer button"
      onClick={[Function]}
    >
      need to create an account?
    </div>
  </div>
</form>
`;

exports[`signup and login 2`] = `
<form
  onSubmit={[Function]}
>
  <h4
    className="mv3"
  >
    Sign Up
  </h4>
  <div
    className="flex flex-column"
  >
    <input
      name="name"
      onChange={[Function]}
      placeholder="Your name"
      type="text"
      value=""
    />
    <input
      name="email"
      onChange={[Function]}
      placeholder="Your email address"
      type="text"
      value=""
    />
    <input
      name="password"
      onChange={[Function]}
      placeholder="Choose a safe password"
      type="password"
      value=""
    />
  </div>
  <div
    className="flex mt3"
  >
    <div
      className="pointer mr2 button"
      onClick={[Function]}
    >
      create account
    </div>
    <div
      className="pointer button"
      onClick={[Function]}
    >
      already have an account?
    </div>
  </div>
</form>
`;
